Add string type

xVin opened the pull request:
targets #35

xVin committed:
Add ws_string interface

Add definitions for string.h

neithernut commented on src/objects/string.h:
```diff
@@ -28,6 +28,158 @@
+
+    struct {
+        size_t charcount;
+        pthread_rwlock_t rwl;
```
Same as for the array type by @matthiasbeyer: one lock is sufficient for the entire type.

neithernut commented on src/objects/string.h:
```diff
@@ -28,6 +28,158 @@
+    } firstbyte; //!< Atomic member for first byte
+
+    struct {
+        char* lastbyte;
```
Can be derived from firstbyte and bytecount.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,140 @@
+
+static bool
+deinit_callback(
+struct ws_object* const
```
Indention missing.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,140 @@
+/* Static Function Implementations*/
+static bool
+deinit_callback(
+struct ws_object* const self
```
Missing indention.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,209 @@
+    ws_object_unlock_read(&other->obj);
+    ws_object_unlock_write(&self->obj);
+    
+    if (self) {
```
The condition should be `self->str`, shouldn't it?

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,209 @@
+    ws_object_lock_write(&self->obj);   
+
+    for (int i = 0; i < others->len; i++) {
+        self = ws_string_cat(self, others[i]);
```
You do not check the return value of `ws_string_cat()` here!

matthiasbeyer replied:
`others[i]` doesn't work. The array type has a function for getting an element at a specific index!

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,209 @@
+    
+    ws_object_unlock_read(&self->obj);
+    
+    if (nstr) {
```
The condition should be `nstr->str`, shouldn't it?

xVin committed:
Add string.c to CMakeLists.txt

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+) { 
+    if (self) {
+        self->charcount = 0; //initialize as empty string
+        self->str = calloc(self->charcount+1, sizeof(*self->str));
```
Spaces around `+` missing.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+            return false;
+        }
+
+        memset(self->str, '\0', sizeof(*self->str));
```
This does not work. You have to `memset` the actual allocated space here, so you should pass `self->charcount + 1` here!

neithernut replied:
Wait...
is the `memset` even neccessary? She used `calloc`...

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+ws_string_dupl(
+    struct ws_string* self
+){
+    ws_object_lock_read(&self->obj);
```
You can move this line to _after: the initialization of the new string (after the `ws_string_new()` call), to save time here.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+){
+    ws_object_lock_read(&self->obj);
+    
+    struct ws_string* nstr = ws_string_new();
```
You do not check the return value, it _could_ be `NULL`

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+    ws_object_lock_read(&self->obj);
+    ws_object_lock_read(&other->obj); //!< @todo Thread-safeness!
+    
+    res = u_strcmp(self->str+offset, other->str);   
```
Nope, `ncmp` should compare the first `n` characters. There must be something like `u_strncmp` or something.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+        return true;
+    }
+
+    else {  
```
The else should be on the line of the closing if bracket, so `} else {` should be on the appropriate line.

Anyways, I assume that the implementation of this is not ready yet.

xVin replied:
Gonna fix this and put everything up for review again!

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+ws_string_raw(
+    struct ws_string* self
+){
+    if (!self->is_utf8) {
```
This is not read-locked.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+
+    ws_object_lock_read(&self->obj);
+
+    output = u_strToUTF8(output, self->charcount, &destL, self->str,
```
What does the `u_strToUTF8()` call?

neithernut replied:
It transforms the internal representation to an utf string and puts it into output. Should be obvious from the name.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,252 @@
+    }
+
+    char* output = calloc(self->charcount, sizeof(char));
+    int32_t destL;
```
Please don't use CamelCase but snake_case. `dest_len` or just `len` should be enough here!

matthiasbeyer commented:
Feel free to do a squash, to clean the history. It's cleaner for you to work with, afterwards.

xVin committed:
Add implementation of ws_string_cat

xVin committed:
Add implementation of ws_string_multicat

xVin committed:
Add implementation of ws_string_dupl

xVin committed:
Add implementation for ws_string_init()

xVin committed:
Add implementation for ws_string_new()

xVin committed:
Add implementation for ws_string_len()

xVin committed:
Add implementation of ws_string_cmp()

xVin committed:
Add implementation of ws_string_ncmp()

xVin committed:
Add implementation of ws_string_substr()

xVin committed:
Add implementation of ws_string_raw()

matthiasbeyer commented:
@xVin Have a look at [this](https://github.com/RaphaelJ/fast_strstr), it _could_ be usefull for the `ws_string_substr()` implementation.

I don't know, but guess, that BSD3C license is compatible with our license, @neithernut , am I right? We could use this code for speeding up some of our string operations (namingly `ws_string_substr()`). We would then do a copy-paste, I assume.

matthiasbeyer commented:
@neithernut , @xVin [this page](https://www.gnu.org/licenses/license-list.html#GPLCompatibleLicenses) says that BSD-3-clause is compatible with GPL.

xVin commented:
I'll have a deeper look at the fast_strstr lib later, but I dunno if it's worth the extra effort - we'd have to convert it to UChars somehow.
For now, I just used ICU's strstr function.

I cleaned up my history (with a little help from my friends), added ws_string_substring and removed the memset in ws_string_init().
Ready for review.

matthiasbeyer commented:
I'm not sure about how much effort this would be, but I guess it wouldn't be that much, as you literally only compare memory. So I guess there is no need to adapt it!

But I'm not that sure about it and would like to hear @neithernut s opinion on this.

I'll review later, too tired right now, so this has to wait until tomorrow!

neithernut commented:
As long as we don't need substrings we should not put in the extra effort.

matthiasbeyer commented:
Please rebase onto the latest `develop` and remove all  `ws_object_unlock_read()` and `ws_object_unlock_write()` calls in favour of the new `ws_object_unlock()` call.

Feel free to do the fixups directly, you don't have to push the single "Fixup: ..." commits!

neithernut commented on src/objects/string.c:
```diff
@@ -25,5 +25,259 @@
+    ws_object_lock_write(&self->obj);
+    ws_object_lock_read(&other->obj); //!< @todo Thread-safeness!
+
+    self->str = u_strcat(self->str, other->str);
```
You need to reallocate `self->str`, so the resulting string is fitting in the memory allocated.

neithernut commented on src/objects/string.c:
```diff
@@ -25,5 +25,259 @@
+    ws_object_lock_write(&self->obj); //!< @todo Thread-safeness!  
+
+    for (unsigned int i = 0; i < others->len; i++) {
+        self->str = u_strcat(self->str, ws_array_get_at(others, i));
```
Same here: you need to reallocate `self->str` beforehand.

matthiasbeyer replied:
And because a `realloc()` call is expensive it would be good to calculate the new length beforehand and not doing a `realloc()` call in every iteration!

xVin replied:
I figured that myself :)

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+        }
+
+        ws_object_init(&self->obj);
+        self->obj.id = &WS_OBJECT_TYPE_ID_ARRAY;
```
This is actually not true, you have the string type here, don't you?

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+struct ws_string*
+ws_string_new(void)
+{
+    struct ws_string* wss = calloc(1, sizeof(struct ws_string));
```
A shortcut would be (for the allocating call) `calloc(1, sizeof(*wss))`

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+    
+    int len = 0;
+    for (unsigned int i = 0; i < others->len; i++) {
+        if (((struct ws_string*) ws_array_get_at(others, i))->str) {
```
That's not good. If the `ws_array_get_at()` call returns NULL, you segfault here!

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+    int len = 0;
+    for (unsigned int i = 0; i < others->len; i++) {
+        if (((struct ws_string*) ws_array_get_at(others, i))->str) {
+            len += ((struct ws_string*) ws_array_get_at(others, i))->charcount;
```
Same as for the line above.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+
+    for (unsigned int i = 0; i < others->len; i++) {
+        self->str = u_strcat(self->str,
+                            ((struct ws_string*)ws_array_get_at(others, i))->str);
```
Here too, this could cause segfaults.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+        return NULL;
+    }
+
+    char* output = calloc(self->charcount, sizeof(char));
```
`sizeof(*output)` - it's the same, but its more maintainable.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+deinit_callback(
+    struct ws_object* const self
+) {
+    free(((struct ws_string*) self)->str);
```
If `self` is `NULL`, you will also segfault here.

xVin replied:
Ok, I modeled this after the deinit() in ws_array, should be fixed there, too, then?

matthiasbeyer replied:
sure.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,273 @@
+    ws_object_unlock(&self->obj);
+    
+    if (self) {
+        return self;
```
You don't need an `if` here, you can just return `self`.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
 
+#include <stdbool.h>
+#include <unicode/ustring.h>
+#include <string.h>
```
Sort them please.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+    .supertype = &WS_OBJECT_TYPE_ID_OBJECT,
+    .typestr = "ws_string",
+    .deinit_callback = deinit_callback,
+};
```
Please add the other entries of the struct and set them to `NULL`.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+    struct ws_string* self,
+    struct ws_array* others
+){
+    if (!self) {
```
`!self || !others` <- how about that? :-)

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+    }
+
+    ws_object_lock_write(&self->obj); //!< @todo Thread-safeness!  
+    
```
You maybe should lock the array object here, too.

But I guess you have to ask @neithernut for this.

xVin replied:
Yeah we should. Itentionally not implemented yet b/c locking solution not ready yet.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+    for (unsigned int i = 0; i < others->len; i++) {
+        struct ws_string* temp = ws_array_get_at(others, i);
+        if (temp && temp->str) {
+                len += temp->charcount;
```
Too much indentation here.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+
+    self->str = realloc(self->str, (len + 1) * sizeof(*self->str));
+
+/* This loop doesn't check for NULL in others, assuming that in the final
```
... in `others` ... (add the backticks around the "others", to make it a code fragment in the documentation.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+        return true;
+    } else {  
+        return false;
+    }
```
Can be shortened to `return !!res;`

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+        return NULL;
+    }
+
+    char* output = calloc(self->charcount, sizeof(*output));
```
Not `self->charcount + 1` (plus one for the trailing zero-byte)?

matthiasbeyer commented:
Gets better with every update!

Btw.: The others ( @neithernut , @TheNeikos , @manuelmessner ) should review too!

xVin committed:
Add ws_string method stubs

xVin commented:
Done with fixing stuff.

matthiasbeyer commented:
@neithernut can you review, too?

matthiasbeyer commented:
@xVin You need to rebase on the latest `develop`, merge conflicts will occur.

neithernut commented on src/objects/string.c:
```diff
@@ -25,5 +25,284 @@
+    .log_callback = NULL,
+    .run_callback = NULL,
+    .hash_callback = NULL,
+};
```
We added a compare-callback lately. Would be nice if you could add an implementation. You may simply pass the `ws_string_cmp` function, but you will probably have to cast the function type.

xVin replied:
Just to clarify: Does that mean declaring ws_string_cmp as a static function (like deinit)?

neithernut replied:
No, it can stay a "public" function, at least in my opinion.

xVin replied:
Okay, but I'm not really sure how to cast the function type then.

neithernut commented:
Ok, looks good. ACK from me.

matthiasbeyer commented on src/objects/string.c:
```diff
@@ -25,5 +25,286 @@
+    .log_callback = NULL,
+    .run_callback = NULL,
+    .hash_callback = NULL,
+    .cmp_callback = (int (*) (struct ws_object const*, struct ws_object const*))
```
Here shouldn't be a cast. The cast should be in the cmp() method. I would prefer a static function which simply does the call to `ws_string_cmp()` over this cast!

If you change the interface definition of the `cmp_callback`, this causes the code to compile anyways, but to crash if the callback gets called while running.

neithernut replied:
This is unneccessary. The cast will fail if the definition changes!

xVin replied:
I agree with neithernut.

matthiasbeyer commented:
Okay, the latest comment was my fault. So ACK from me.

neithernut commented:
Happy squashing!

matthiasbeyer commented:
No squashing required here, at least in my opinion.

neithernut commented:
k, then ACK.

matthiasbeyer commented:
@neithernut go ahead and merge! :-P
